### PR TITLE
Importer fails on missing label instead of missing table.

### PR DESCRIPTION
**Pull request by SamGerber-zz**
#### Context:

When we instantiate a new `LiveFixtures::Import`, one of the arguments we must provide is an `insert_order`.

We use labels as placeholders for foreign_keys, so that records that are associated in the DB can continue to reference each other after being exported. On import, we replace these labels with the newly imported record's new ID, so the imported records will be properly associated again.

```YML
# In users.yml
users_1:
    posts: 'posts_1,posts_2'

# In posts.yml
posts_1:
    ...

posts_2:
    ...
```

The labels `users_1`, `posts_1`, and `posts_2` will be replaced with the fixture's new ID.

In order for this all to work properly, we have to import any fixture that is referenced BEFORE we try to import any fixture that references it. This way, when we import users from the example above, we'll have already imported posts and we'll know the IDs we should use to replace the labels `posts_1` and `posts_2`.

So, in the past we've been raising an error when we can't find one of the files listed in `insert_order`. This probably was meant to help keep us from forgetting to import posts.

This is a bit of a drag, because if we export a set of fixtures for one of users who doesn't have any posts, no posts.yml file will be generated, and the import will fail before it begins, even though the fixtures are valid.

Instead, we now raise a `LiveFixtures::MissingReferenceError` if we attempt to replace a label and cannot find a matching ID.

The error should help point you in the direction of what went wrong:

>Unable to find ID for model referenced by label not_the_right_label while importing Trogolodyte from trogolodytes.yml. Perhaps it isn't included in these fixtures or it is too late in the insert_order and has not yet been imported.

#### Github Issue:
Outfoxes: #13
 
 #### Merging this PR
 
 - [ ] Double check that all reviewers' comments were addressed or acknowledged.
    - Comments about future refactoring should end up as TODOs in the code.
 - [ ] **SQUASH** and Merge the PR.
 - [ ] Delete the branch.

### CODE REVIEW

**Comment by jleven** [comment]
@SamGerber - LiveFixtures was designed to be safe to use against a production db. That's why we have validations before any data is imported -- when things mess up, we can't just wipe the db. Maybe skipping that initial validation can be a parameter or environment setting instead?

**Comment by SamGerber-zz** [comment]
When what things mess up?

**Comment by SamGerber-zz** [comment]
Oh, I think I see what you mean. Isn't the whole import wrapped in a transaction, though, so if it messes up it will roll back?
https://github.com/NoRedInk/live_fixtures/blob/master/lib/live_fixtures/import.rb#L46-L63

**Comment by jleven** [comment]
Oh, I definitely forgot about that. Good thinking past-Josh! I guess, nevermind then?

File: lib/live_fixtures/import/fixtures.rb [hunk 1]
```diff
@@ -62,8 +67,15 @@ def each_table_row_with_label
       join_table_rows.each do |table_name, rows|
         rows.each do |targets:, association:, label:|
           targets.each do |target|
-            assoc_fk = @label_to_id[target] || target
-            row = { association.foreign_key             => @label_to_id[label],
+            # This song and dance is so we can specify HABTM like so:
+            # dogs: 12, 32, 144 instead of dogs: dogs_12, dogs_32, dogs_144
```
Comment (jleven) [inline lib/live_fixtures/import/fixtures.rb:71]: LiveFixtures works with yml files that were created by hand as well. There are no requirements on the format of the label for each record.

File: lib/live_fixtures/import/fixtures.rb [hunk 2]
```diff
@@ -77,6 +89,33 @@ def model_connection
...
+    # Checks if the string looks like a label referencing the specified table.
+    # @param label_to_check [String] the string that might be a label.
+    # @param table_name [String] the table the label might reference.
+    # @return [Boolean] whether the string is a label for the table.
+    # @see LiveFixtures::Export::Reference
+    # @see LiveFixtures::Export::Fixture#yml_value
+    def is_label_for_table?(label_to_check, table_name)
+      label_prefix = table_name + "_"
+      label_to_check.starts_with? label_prefix
+    end
```
Comment (jleven) [inline lib/live_fixtures/import/fixtures.rb:117]: Can you help me understand why we are doing this? I don't know that we need to assume labels are in a particular format.

**Review by jleven — COMMENTED** [review]
I'm a little unsure about this change. I have to admit I'm less familiar with the code than you are now @SamGerber. Still, if the goal is only to have an error when the foreign-key-label can't be found, I don't see the advantage to adding much of this code. It appears to enforce a particular format for the labels in the yml when there was no such enforcement before. Can you help me understand what the advantage is of adding this requirement?

File: lib/live_fixtures/import/fixtures.rb (hunk 1 above)
Comment (SamGerber) [inline lib/live_fixtures/import/fixtures.rb]: I see. I saw that AR fixtures worked that way, but didn't realize live_fixtures had that capability as well.

**Comment by SamGerber-zz** [comment]
Thanks for the review Josh! You definitely still have more context on this than I do! I didn't even consider that labels could take other forms.

In my use of the gem already, it's been pretty valuable to have it fail when it cannot find a label; it's tipped me off to malformed exports twice. It seems like we could accomplish this without requiring a format for labels, because we can deduce whether a value will be a label or a foreign key based on the attribute name. We'd expect `user` to point to a label, while `user_id` ought to point to a foreign key. It looks like that's how you've handled things in [maybe_convert_association_to_foreign_key](https://github.com/NoRedInk/live_fixtures/blob/master/lib/live_fixtures/import/fixtures.rb#L93).

(This does mean we couldn't have an association called user_id, but I think that's an okay assumption to make.)

Where I thought things broke a bit was in allowing the flexibility for HABTM associations.

By allowing either:
```
dogs: 1, 2, 3
dogs: dogs_1,dogs_2,dogs_3
```
It seems like we cannot tell whether the value is a list of labels or a list of foreign keys.

This flexibility is nice, but I think it also means if we had a label `dogs_2` whose record we never exported, we'd get nil when we try to look it up in @label_to_id. For HABTM, that means it would fall back on the target, itself 'dogs_2' and try to use that as the foreign_key. There's a good chance that would blow up in the DB.

For belongs_to associations, though, it would return nil, insert NULL, and we'd be none the wiser.

What if we enforce the attribute name in the HABTM case, too?
```
dog_ids: 1, 2, 3
dogs: dogs_1,dogs_2,dogs_3
```

Maybe there's a DSL for activerecord fixtures that this follows and we'd break convention in doing that, but I really do think there's value in having the integrity check.

**Comment by SamGerber-zz** [comment]
Also, I like the idea of a parameter, maybe something like, in case of missing label, 
 - skip the record's insertion,
 - replace with nil, insert the record, and warn
 - fail

**Comment by SamGerber-zz** [comment]
@jleven , I'd love to pick your brain on this if you have time sometime!

**Comment by SamGerber-zz** [comment]
Closing in favor of: https://github.com/NoRedInk/live_fixtures/pull/23

**Pull request closed without merging**